### DISCUSSION TITLE
WSL's design should be changed

### DISCUSSION THREAD

**Discussion by wzis** [microsoft/WSL#9304] [Ideas]
Current WSL is not shareable among the accounts on the machine, that's a big problem. By redesigning it to only allow admin account to create the WSL instances and assign a name for each instance, then letting users on the same machine use ssh to an instance and an account, it will become much more useful.  Also by using an environment variable to specify the default instance and default account on the instance, and requiring that the default account login needs be setup using the ssh public key method, it will allow the PC user to use wsl to run Linux command using the default instance and default account. This way, WSL's usage will be more natural and the instance will behave more naturally.

**Comment by simonSilverse** [comment]
<pre>
Based on documentation:
	The WSL uses Hyper-V architecture for virtualization and may not work together with other virtual Machine (VM).
Based on own findings:
	Only 1 VM per Windows user account (WUA) is used with primary root process starting from distro 'init' binary.
	Each seperate shell consists of a 'SessionLeader' and a 'Relay', both originating from 'init' binary.
	When no session exist the VM will shutdown.
	WSL is installed as a system wide program.
	WSL starts with its own VM under the hood for the current WUA.
	It seems like the distro/VM cannot be shared across WUA.
Unverified possibilities:
	A SSH server can be adopted to enable access for multiple users outside the context of the WUA.
	However, that does not provide flexibility in number of instance management per WUA.
	A host machine with SSH server plus a version of WSL with enhanced VM management may be the desirable design.

The current package name WSL sounds fine and I think any architectural redesign should be done by further breaking down into:
	!) WSLM 	// M for machine
	2) WSLD		// D for daemon
	3) WSLC		// C for client
The client interacts with the daemon which manages the VM instance or container.
That is as close as having a Windows self implementation of Docker...
</pre>

**Reply by wzis** [reply]
When WSL VM is only created by the admin, it can be started by the Windows system service. When Windows accounts need to bind to the Linux accounts, use the ssh public key and define the binding as a specific environment variable value. When the public key is set up properly, it should allow wsl to run commands directly as the current way. That way, the WSL is more useful and natural.

**Comment by blakeduffey** [comment]
This would not be a desirable change.  In a corporate environment, users often don't have local admin privileges.

**Reply by wzis** [reply]
In corporate env, users should ask admins to create the WSL for them to use, but they should not get root access by default.
